protocol/messenger_communities.go

```diff
@@ -1947,6 +1974,95 @@ func (m *Messenger) CreateCommunity(request *requests.CreateCommunity, createDef
 	return response, nil
 }
 
+func (m *Messenger) SetCommunityShard(request *requests.SetCommunityShard) (*MessengerResponse, error) {
+	if err := request.Validate(); err != nil {
+		return nil, err
+	}
+
+	community, err := m.communitiesManager.SetShard(request.CommunityID, request.Shard)
+	if err != nil {
+		return nil, err
+	}
+
+	var topicPrivKey *ecdsa.PrivateKey
+	if request.PrivateKey != nil {
+		topicPrivKey, err = crypto.ToECDSA(*request.PrivateKey)
+	} else {
+		topicPrivKey, err = crypto.GenerateKey()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.UpdateCommunityFilters(community, topicPrivKey)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.SendCommunityShardKey(community, community.GetMemberPubkeys())
+	if err != nil {
+		return nil, err
+	}
+
+	response := &MessengerResponse{}
+	response.AddProtectedTopic(community.PubsubTopic(), topicPrivKey)
+
+	return response, nil
+}
+
+func (m *Messenger) UpdateCommunityFilters(community *communities.Community, privKey *ecdsa.PrivateKey) error {
+	if m.transport.WakuVersion() == 2 {
+		if err := m.transport.StorePubsubTopicKey(community.PubsubTopic(), privKey); err != nil {
+			return err
+		}
+	}
+
+	publicFiltersToInit := make([]transport.FiltersToInitialize, 0, len(community.DefaultFilters())+len(community.Chats()))
+
+	for _, df := range community.DefaultFilters() {
+		_, err := m.transport.RemoveFilterByChatID(df.ChatID)
+		if err != nil {
+			return err
+		}
+		publicFiltersToInit = append(publicFiltersToInit, df)
+	}
+
+	for chatID := range community.Chats() {
+		communityChatID := community.IDString() + chatID
+		_, err := m.transport.RemoveFilterByChatID(communityChatID)
+		if err != nil {
+			return err
+		}
+		publicFiltersToInit = append(publicFiltersToInit, transport.FiltersToInitialize{ChatID: communityChatID, PubsubTopic: community.PubsubTopic()})
+	}
+
+	_, err := m.transport.InitPublicFilters(publicFiltersToInit)
```

Comment:
maybe we could refactor this function? it seems the same logic could be used somewhere else?